lib/ecto/repo/schema.ex

```diff
@@ -75,46 +75,66 @@ defmodule Ecto.Repo.Schema do
     end
   end
 
-  defp extract_header_and_fields(rows, schema, dumper, autogen_id, adapter, placeholder_map) do
-    mapper = init_mapper(schema, dumper, adapter, placeholder_map)
+  defp extract_header_and_fields(rows, schema, dumper, autogen_id, placeholder_map, adapter) do
+    mapper = init_mapper(schema, dumper, adapter)
 
-    {rows, {header, has_query?}} =
-      Enum.map_reduce(rows, {%{}, false}, fn fields, acc ->
-        {fields, {header, has_query?}} = Enum.map_reduce(fields, acc, mapper)
+    {rows, {header, has_query?, placeholder_map}} =
+      Enum.map_reduce(rows, {%{}, false, placeholder_map}, fn fields, acc ->
+        {fields, {header, has_query?, placeholder_map}} = Enum.map_reduce(fields, acc, mapper)
         {fields, header} = autogenerate_id(autogen_id, fields, header, adapter)
-        {fields, {header, has_query?}}
+        {fields, {header, has_query?, placeholder_map}}
       end)
 
+    placeholder_map = placeholder_map
+    |> Enum.map(fn {k, {_, v}} -> {k, v} end)
+    |> Map.new
+
     if has_query? do
       rows = plan_query_in_rows(rows, header, adapter)
-      {rows, header}
+      {rows, header, placeholder_map}
     else
-      {rows, header}
+      {rows, header, placeholder_map}
     end
   end
 
-  defp init_mapper(nil, _dumper, _adapter, _placeholder_map) do
-    fn {field, _} = tuple, {header, has_query?} ->
-      {tuple, {Map.put(header, field, true), has_query?}}
+  defp init_mapper(nil, _dumper, _adapter) do
+    fn {field, _} = tuple, {header, has_query?, placeholder_map} ->
+      {tuple, {Map.put(header, field, true), has_query?, placeholder_map}}
     end
   end
 
-  defp init_mapper(schema, dumper, adapter, placeholder_map) do
-    fn {field, value}, {header, has_query?} ->
+  defp init_mapper(schema, dumper, adapter) do
+    fn {field, value}, {header, has_query?, placeholder_map} ->
       case dumper do
         %{^field => {source, type}} ->
           case value do
             %Ecto.Query{} = query ->
-              {{source, query}, {Map.put(header, source, true), true}}
+              {{source, query}, {Map.put(header, source, true), true, placeholder_map}}
 
             {:placeholder, placeholder_key} ->
-              replaced_value = Map.fetch!(placeholder_map, placeholder_key)
-              value = dump_field!(:insert_all, schema, field, type, replaced_value, adapter)
-              {{source, value}, {Map.put(header, source, true), has_query?}}
+              placeholder_map = case placeholder_map do
+                %{^placeholder_key => {^type, _}} = map ->
```

Comment:
As it is implemented today, you can run into issues if for some reason there is a type that actually supports two element tuple as arguments. For example, imagine someone does this:

    insert_all([...], placeholders: %{foo: {:my_custom_type, :value}})

So we would need to keep the types and the values separately. I am also thinking that when we do this pass, we should already return the placeholders as a list and the entries in the insert all fields should already be `{:placeholder, index}`. So when we call the adapter, the simply pass a list with all placeholders.

So to be clear, we will call:

    adapter.insert_all(adapter_meta, schema_meta, Map.keys(header), rows, on_conflict, return_sources, placeholders, opts)

where placeholders is already a list:

    [Date.today, dumped_foo, dumped_bar, ...]

The rows will have entries like this:

    {:placeholder, 0}, {:placeholder, 1}, ... # starting from zero

Then in the Ecto.Adapters.SQL code you won't need to change unzip_inserts anymore, you will simply do this:

    placeholders ++ Enum.reverse(params) ++ on_conflict_values

And in Connection.insert you will receive all placeholders and you can compute the relative starting index by calling `length(placeholders)`.

This approach will make sure most of the placeholder logic is handled here, which is the shared code, and the placeholders logic in adapters is kept to a minimum.